src/github.com/matrix-org/dendrite/roomserver/storage/sql.go

```diff
@@ -333,6 +340,13 @@ CREATE TABLE IF NOT EXISTS events (
     -- Local numeric ID for the state_key of the event
     -- This is 0 if the event is not a state event.
     event_state_key_nid BIGINT NOT NULL,
+	-- Local numeric ID for the state at the event.
+    -- This is 0 if we don't know the state at the event.
+    -- If the state is not 0 this this event is part of the contiguous
+    -- part of the event graph
+    -- Since many different events can have the same state we store the
+    -- state into a separate state table and refer to it by numeric ID.
+    state_nid bigint NOT NULL DEFAULT 0
```

Comment:
Naming: I don't like the name `state_nid` because it's not obvious what it conceptually represents in Matrix-land: a state event, state tuple, state key? Based on your comments, it looks like this represents the room state at a particular point in time, so perhaps `state_snapshot_id`?